### ISSUE TITLE
Cython grammar support

### ISSUE THREAD

**Issue by ambv** [psf/black#359]
We'd like to be able to format .pyx, .pxd, and .pxi files, too.

**Comment by ambv** [comment]
Hi, @pablogsal.

**Comment by GrigoriosGiann** [comment]
Hello Łukasz!

I have been working on extending black for Cython.
Here is my approach:
Use a different grammar file, for the Cython grammar.
Start with the Python grammar and add Cython grammar rules to it one by one.
I have added most "simple" rules to the grammar so far, and I'm adding formatting rules for the blindingly obvious cases.

Here are some issues I faced in my endeavor:
- As a result of using a different grammar file, there is a symbol mismatch (the same symbol may be represented by a different integer in the Python grammar and Cython grammar). This requires functions and top level variables using `type_repr` and `syms` to be parameterized with regards to the grammar's symbols. This does not affect tokens. (this may require encapsulating some of the global state)
- Because of the way `pgen` works right now, string literals in the grammar cannot match `NAME` in the grammar. This is not that big of an issue in some cases and can be solved by disabling the keywords, however in other cases it's much harder as it conflicts with other rules. (like with the new expression: `new = 5` vs `new vector[int]()`) I plan on deferring this issue for later.

Also I've expanded the tokenizer to accept integer literals ending in an arbitrary amount of (`l`, `L`, `u`, `U`), which affects affects python parsing as well. (single (`l`/`L`)-suffixed integer literals were already tokenized, even though they are invalid in Python3.)

Once I gather a sizable amount of rules, I'll let you know so you can inform me of your preferred formatting rules
I'd love to hear any suggestions/recommendations you have with regards to anything I've implemented so far, or any preference on how to proceed on the mentioned issues.

**Comment by pablogsal** [comment]
[Just as a clarification, @GrigoriosGiann is working with us in the Cython support as part of his internship]

**Comment by jakirkham** [comment]
@pablogsal @GrigoriosGiann, do you have something that you would like people to try/provide feedback on?

**Comment by Synss** [comment]
Are there any news on this? Even if you only have C syntax (no C++) or only `.pxd` or something, it would be really interesting to publish...

**Comment by pablogsal** [comment]
> Are there any news on this?

Yeah, we will soon do a PR to iterate over :). I was caught in many other responsibilities and could not finish some required steps until now.

**Comment by RyanHope** [comment]
```
error: cannot format filename.pyx: Cannot parse: 11:13: from cpython cimport array
```

im getting this on latest dev version

**Comment by jlucier** [comment]
Is there any more update on this? I'm running into issues still with formatting Cython. Any tips @pablogsal?

**Comment by ArseniiDunaevQC** [comment]
I have a small update on the issue since I was working on it a bit for the last time. Basically I have rebased changes made by @pablogsal (https://github.com/pablogsal/black/tree/cython - hope you don't mind me taking it) on top of the current `master` branch and tried to add absent Cython features/rules, and to fix some existing bugs (you can check out my WIP-branch at https://github.com/ArseniiDunaevQC/black/tree/cython-dev). 
First of all, it could be already (limitedly) used for some of the basic `.pyx` files. Here I would for sure be happy to hear any suggestions/recommendations as well as preferred formatting rules.

There are still some other issues that hinder further development:
- As @grigoriosgiann mentioned above, there are difficulties with matching grammar string literals with `NAME`s. The hardest things for the moment are Python enumerations (`Enum` class) and using `new` as variable name.
- It would be great to have support for Cython compile-time constants and conditional compilation using `DEF`, `IF` etc. keywords (see [docs](https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#conditional-compilation)) and `include`.

**Comment by pablogsal** [comment]
> Hi all!
> 
> I have a small update on the issue since I was working on it a bit in the last time. Basically I have rebased changes made by @pablogsal (https://github.com/pablogsal/black/tree/cython - hope you don't mind me taking it) 

The actual work in progress that we had does not love there but in https://github.com/bloomberg/black (branch `cython_support`). If you are going to base your work on top of this, please talk with @grigoriosgiann  first.

**Comment by TheRealBecks** [comment]
What is needed to get Black working with Cython? Is there any way to push that integration? :)

**Comment by cooperlees** [comment]
> What is needed to get Black working with Cython? Is there any way to push that integration? :)

https://github.com/bloomberg/black/tree/cython_support - Take this, rebase, retest etc. and get a PR up ... We would need pretty comprehensive tests to accept tho - But I am sure we can find people to help with the testing.

It probably needs a pick and pull from the branch rewrite as we've totally refactored out of one __init__.py / single py file to lots of modules. This would be cool to support.

**Comment by RyanHope** [comment]
I use cython extensively in my projects so I can do a lot of testing, but I
don't have time to dive into someone else's project right now and rewrite
code.

**Comment by JelleZijlstra** [comment]
That branch is based on augmenting blib2to3 to parse Cython. But that's a problem because we eventually want to replace blib2to3 with a PEG-based parser so that we can keep parsing newer versions of Python.

**Comment by kennylajara** [comment]
Hope to see the release soon

**Comment by TheRealBecks** [comment]
So the Cython integration is stuck?

**Comment by sabard** [comment]
Any progress on this? I'd be happy to take a look at pushing forward @grigoriosgiann's solution sometime this winter if not.

**Comment by JelleZijlstra** [comment]
No progress. We have a vague intention to switch to a LibCST parser at some point, but as far as I know there's no concrete work being done at the moment.

**Comment by MuellerSeb** [comment]
For the record: There is at least cython-lint to gain some formatting standard for cython: https://pypi.org/project/cython-lint/

**Comment by Vizonex** [comment]
I would be willing to try and go in and add cython support for this library since [ruff doesn't have as much of a desire for it yet](https://github.com/astral-sh/ruff/issues/10250) and at least trying to get cython support implemented here would be less intense for me and it may lead to healthy friendly competition over on ruff's end. My only question would be how we setup a separate AST for Cython but also modify the blib2to3 code to incorporate Cython as a secondary grammar set. I do know about some existing forks with these additions but the ones that I've stumbled across seem to be a bit outdated. Rewriting the any previous ASTs altogether might be necessary to accommodate for any newer features that may have been added to Cython.

**Comment by JelleZijlstra** [comment]
Possibly the best approach would be to have a separate Grammar file for Cython, presumably with the existing Python grammar plus some Cython-specific rules. Then we can expand the formatter to deal with those extra grammar nodes. I have a toy project at https://github.com/JelleZijlstra/lib2toast that can work kind of like this: it lets you use a custom grammar with blib2to3.

**Comment by dellelce** [comment]
After 8 years it is better to give up and surrender.
